src/engraving/libmscore/harppedaldiagram.cpp

```diff
@@ -0,0 +1,370 @@
+            while (prevState == _pedalState) {
+                // if states are the same but previous is a diagram write whole config out
+                if (prevDiagram->isDiagram()) {
+                    prevState = initState;
+                    break;
+                }
+                // Check if there's a diagram before.  If not, start of score and use init state
+                prevDiagram = part()->prevHarpDiagram(prevDiagram->segment()->tick());
+                if (prevDiagram != nullptr) {
+                    prevState = prevDiagram->getPedalState();
+                } else {
+                    // set to init state and break loop
+                    prevState = initState;
+                    break;
+                }
+            }
+        } else {
+            prevState = initState;
+        }
+
+        String topLine, bottomLine;
+
+        for (int idx = 0; idx < _pedalState.size(); idx++) {
+            if (_pedalState[idx] != prevState[idx]) {
+                String strName = harpStringTypeToString(HarpStringType(idx));
+                switch (_pedalState[idx]) {
+                case PedalPosition::FLAT:
+                    strName.append(u"<sym>csymAccidentalFlat</sym> ");
+                    break;
+                case PedalPosition::NATURAL:
+                    strName.append(u"<sym>csymAccidentalNatural</sym> ");
+                    break;
+                case PedalPosition::SHARP:
+                    strName.append(u"<sym>csymAccidentalSharp</sym> ");
+                    break;
+                case PedalPosition::UNSET:
+                    break;
+                }
+                if (idx < SEPARATOR_IDX) {
+                    bottomLine.append(strName);
+                } else {
+                    topLine.append(strName);
+                }
+            }
+        }
+
+        if (topLine.size() == 0) {
+            diagram.append(bottomLine);
+        } else {
+            diagram.append(topLine + u"\n" + bottomLine);
+        }
+    }
+
+    return diagram;
+}
+
+void HarpPedalDiagram::updateDiagramText()
+{
+    undoChangeProperty(Pid::TEXT, createDiagramText(), PropertyFlags::STYLED);
+}
+
+bool HarpPedalDiagram::isPitchPlayable(int pitch)
+{
+    return _playablePitches.find(pitch % 12) != _playablePitches.cend();
+}
+
+void HarpPedalDiagram::read(XmlReader& xml)
+{
+    rw400::TRead::read(this, xml, *xml.context());
+}
+
+void HarpPedalDiagram::write(XmlWriter& xml) const
+{
+    if (!xml.context()->canWrite(this)) {
+        return;
+    }
+    xml.startElement(this);
+    writeProperty(xml, Pid::HARP_IS_DIAGRAM);
+
+    // Write vector of harp strings.  Order is always D, C, B, E, F, G, A
+    xml.startElement("pedalState");
+    for (int idx = 0; idx < _pedalState.size(); idx++) {
+        xml.tag("string", { { "name", idx } }, static_cast<int>(_pedalState[idx]));
+    }
+    xml.endElement();
+
+    TextBase::writeProperties(xml);
+    xml.endElement();
+}
```

Comment:
You'll need to rebase once more; the changes are the following:
- The `read` method has been removed definitively from all `EngravingItem` classes;
- The `write` method too; its functionality has been moved to the `TWrite` class;
- The headers `"rw/xml.h"` and `"rw/400/tread.h"` are no longer being included in files for `EngravingItem` classes.